### PR TITLE
Add `func DictObject`

### PR DESCRIPTION
**Pull request by MayCXC**
Allows fields to be used to construct `ObjectMarshaler`s that can then be used with other functions like `func Objects`. Link to issue: https://github.com/uber-go/zap/issues/1458

### CODE REVIEW

**Comment by abhinav** [comment]
Thanks for the PR, @MayCXC.
This change makes sense to me.
Would you mind adding some documentation and tests for this?

**Comment by MayCXC** [comment]
> Thanks for the PR, @MayCXC. This change makes sense to me. Would you mind adding some documentation and tests for this?

you got it, I added doc comments & tests.

File: example_test.go [hunk 1]
```diff
@@ -388,6 +388,60 @@ func ExampleObjects() {
...
+		panic(err)
+	}
+	d2, err := time.ParseDuration("79ms")
+	if(err != nil) {
+		panic(err)
+	}
+	d3, err := time.ParseDuration("57ms")
+	if(err != nil) {
+		panic(err)
+	}
```
Comment (abhinav) [inline example_test.go:418]: can we just use time.Duration values directly here?

```suggestion
	d1 := 68 * time.Millisecond
	d2 := 79 * time.Millisecond
	d3 := 57 * time.Millisecond
```

File: field.go [hunk 2]
```diff
@@ -431,6 +431,12 @@ func (d dictObject) MarshalLogObject(enc zapcore.ObjectEncoder) error {
 	return nil
 }
 
+// DictObject exposes the [zapcore.ObjectMarshaler] for an array of [Field] to
+// use with functions like [Object] and [Objects].
```
Comment (abhinav) [inline field.go:435]: I would reword this a bit. Object/Objects aren't the only uses of this.

```suggestion
// DictObject constructs a [zapcore.ObjectMarshaler] with the given list of fields.
// The resulting object may be used as input to [Object], [Objects],
// or anywhere else that an object is expected.
```

File: zapcore/field_test.go [hunk 3]
```diff
@@ -320,6 +320,16 @@ func TestEquals(t *testing.T) {
 			b:    zap.Dict("k", zap.String("a", "d")),
 			want: false,
 		},
+		{
+			a:    zap.Object("k", zap.DictObject(zap.String("a", "b"))),
+			b:    zap.Object("k", zap.DictObject(zap.String("a", "b"))),
+			want: true,
+		},
+		{
+			a:    zap.Object("k", zap.DictObject(zap.String("a", "b"))),
```
Comment (abhinav) [inline zapcore/field_test.go:329]: The example test is good, and this equality check makes sense.
Would you also mind adding a test similar to [TestDict](https://github.com/uber-go/zap/blob/3f27eb9c787ea9b2ddfc1f707ef291c9070655fe/field_test.go#L294) that asserts the shape of the resulting JSON-ey object.

File: example_test.go (hunk 1 above)
Comment (MayCXC) [reply example_test.go]: yep, this looks much better

File: field.go (hunk 2 above)
Comment (MayCXC) [reply field.go]: I reworded it a bit based on this feedback

File: zapcore/field_test.go (hunk 3 above)
Comment (MayCXC) [reply zapcore/field_test.go:329]: sounds good, I put one here: https://github.com/uber-go/zap/pull/1460/files#diff-218ab55776e43f87dc44a4b65eebeff94670cebeb5dcf80ca3143b5577b50e7dR318

**Review by abhinav — APPROVED** [review]
LGTM!
Will wait for an Uber maintainer to also stamp before merging.

CC @sywhang @r-hang @JacobOaks

**Review by sywhang — APPROVED** [review]
this looks good to me. Thanks for the PR @MayCXC !
